## Упражнения по библиотеке Numpy

In [1]:
try:
    import numpy as np
except ModuleNotFoundError:
    %pip install numpy
    import numpy as np
import math
import random
try: 
    import scipy
except ModuleNotFoundError:
    %pip install scipy
    import scipy
from datetime import datetime

try:
    import plotly.graph_objects as go
except ModuleNotFoundError:
    %pip install plotly
    import plotly.graph_objects as go


**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
arr = np.random.randint(0, 10, 10)
print(arr)
print(np.vectorize(lambda x: -1*x if x >3 and x< 8 else x)(arr))

[7 0 4 7 6 3 0 0 9 4]
[-7  0 -4 -7 -6  3  0  0  9 -4]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr = np.random.randint(0, 10, 10)
print(arr)
print(np.vectorize(lambda x,y: 0 if x == y else x)(arr,arr.max()))

[9 9 0 6 6 0 4 5 2 0]
[0 0 0 6 6 0 4 5 2 0]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
arr, arr2 = np.random.randint(0, 10, 10),np.random.randint(0, 10, (10,1))
print(arr,arr2,sep='\n')
arr*arr2

[4 3 0 1 3 0 1 4 2 8]
[[7]
 [5]
 [0]
 [1]
 [8]
 [7]
 [8]
 [6]
 [6]
 [1]]


array([[28, 21,  0,  7, 21,  0,  7, 28, 14, 56],
       [20, 15,  0,  5, 15,  0,  5, 20, 10, 40],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  3,  0,  1,  3,  0,  1,  4,  2,  8],
       [32, 24,  0,  8, 24,  0,  8, 32, 16, 64],
       [28, 21,  0,  7, 21,  0,  7, 28, 14, 56],
       [32, 24,  0,  8, 24,  0,  8, 32, 16, 64],
       [24, 18,  0,  6, 18,  0,  6, 24, 12, 48],
       [24, 18,  0,  6, 18,  0,  6, 24, 12, 48],
       [ 4,  3,  0,  1,  3,  0,  1,  4,  2,  8]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
arr, arr2= np.random.randint(0, 10, (8, 3)),np.random.randint(0, 10, (2, 2))
print(arr,arr2,sep='\n')

def list_bool_or(arr,arr2):
    res1 = res2 = False
    for x in arr:
        res1 = x in arr2[0] or res1
        res2 = x in arr2[1] or res2
    return res1 and res2

np.array([x for x in arr if list_bool_or(x, arr2)])

[[0 3 1]
 [4 4 0]
 [1 9 8]
 [7 9 8]
 [5 5 3]
 [1 3 6]
 [0 6 0]
 [1 1 4]]
[[9 2]
 [1 3]]


array([[1, 9, 8]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
arr = np.random.randint(0, 5, (10, 2))
print(arr)

def list_check(arr):
    first = arr[0]
    for i in arr[1:]:
        if first != i:
            return True
    return False

np.array([x for x in arr if list_check(x)])

[[1 3]
 [1 0]
 [2 4]
 [2 0]
 [2 3]
 [2 1]
 [1 0]
 [0 2]
 [2 0]
 [1 2]]


array([[1, 3],
       [1, 0],
       [2, 4],
       [2, 0],
       [2, 3],
       [2, 1],
       [1, 0],
       [0, 2],
       [2, 0],
       [1, 2]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
arr = np.random.randint(0, 5, (10, 2))
print(arr)
np.unique(arr, axis=0)

[[4 0]
 [3 4]
 [1 1]
 [0 1]
 [4 2]
 [2 4]
 [4 0]
 [1 1]
 [2 2]
 [2 1]]


array([[0, 1],
       [1, 1],
       [2, 1],
       [2, 2],
       [2, 4],
       [3, 4],
       [4, 0],
       [4, 2]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
l = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
res = 1
for i in range(min(len(l),len(l[0]))):
    if l[i][i] != 0:
        res*= l[i][i]
res

3

In [9]:
arr = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
np.multiply.reduce(math.prod(arr.diagonal()[arr.diagonal()!=0]))

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [10]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
set(x)==set(y)

True

In [11]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
np.array_equal(np.unique(x),np.unique(y))

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [12]:
l = [6, 2, 0, 3, 0, 0, 5, 7, 0]
res = 0
for i in range(1,len(l)):
    if l[i-1] == 0:
        res = max(l[i], res)
res

5

In [13]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

np.apply_along_axis(lambda x: x[1] if x[0] ==0 else -1,1,np.vstack((np.roll(x,1)[1:],x[1:])).T).max()

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [14]:
l = [2, 2, 2, 3, 3, 3, 5]
var, count = [], []
for i in set(l):
    var.append(i)
    count.append(l.count(i))
print(var,count)

[2, 3, 5] [3, 3, 1]


In [15]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [16]:
# plain python
x, y = np.random.randint(0, 10, (4, 4)).tolist() , np.random.randint(0, 10, (4, 4)).tolist()
print(x,y,sep='\n')

def matrix_of_distance_plaint(x,y):

    def distance(x,y):
        res = 0
        for i in range(len(x)):
            res+= (x[i] - y[i])**2
        return res**(0.5)
    
    res = []
    for ix in x:
        row = []
        for iy in y:
            row.append(distance(ix,iy))
        res.append(row)
    return res

matrix_of_distance_plaint(x,y)

[[5, 6, 4, 3], [2, 6, 9, 5], [6, 2, 5, 6], [8, 7, 4, 4]]
[[2, 9, 1, 5], [0, 8, 0, 9], [5, 6, 6, 8], [3, 4, 4, 3]]


[[5.5677643628300215, 9.0, 5.385164807134504, 2.8284271247461903],
 [8.54400374531753, 10.246950765959598, 5.196152422706632, 5.830951894845301],
 [9.055385138137417, 10.295630140987, 4.69041575982343, 4.795831523312719],
 [7.0710678118654755, 10.295630140987, 5.477225575051661, 5.916079783099616]]

In [17]:
# numpy
x, y = np.random.randint(0, 10, (4, 4)) ,np.random.randint(0, 10, (4, 4))
print(x,y,sep='\n')

def matrix_of_distance_numpy(x,y):
    dist = lambda x,y: np.sqrt(np.square(x-y).sum())
    join = lambda f: np.apply_along_axis(lambda x,y: dist(x,y),1,x,**{'y':f})
    return np.apply_along_axis(join,1,y)

matrix_of_distance_numpy(x,y)

[[0 2 7 7]
 [9 0 5 1]
 [5 4 4 6]
 [8 6 0 4]]
[[5 0 2 3]
 [8 4 6 4]
 [9 1 4 1]
 [0 0 2 2]]


array([[ 8.36660027,  5.38516481,  5.38516481,  7.07106781],
       [ 8.83176087,  5.19615242,  4.12310563,  6.32455532],
       [11.26942767,  1.41421356,  7.07106781,  7.14142843],
       [ 7.34846923,  9.53939201,  7.81024968, 10.39230485]])

In [18]:
# scipy.spatial.distance.cdist
x, y = np.random.randint(0, 10, (4, 4)) ,np.random.randint(0, 10, (4, 4))
print(x,y,sep='\n')

start = datetime.now()
r = scipy.spatial.distance.cdist(x,y)
print(f'exec time = {datetime.now() - start}')

print(r)

[[4 3 5 5]
 [3 1 1 1]
 [9 1 3 9]
 [1 2 0 8]]
[[1 6 4 4]
 [2 4 3 2]
 [0 9 5 6]
 [9 2 8 8]]
exec time = 0:00:00.049872
[[ 4.47213595  4.24264069  7.28010989  6.63324958]
 [ 6.8556546   3.87298335 10.67707825 11.61895004]
 [10.72380529 10.34408043 12.56980509  5.19615242]
 [ 6.92820323  7.07106781  8.88819442 11.3137085 ]]


In [19]:
measurements = [[],[],[],[]]
times = 100
for n in range(4,11):
    x, y = np.random.randint(0, 10, (n, 4)) ,np.random.randint(0, 10, (n, 4))
    measurements[0].append(n)
    res = [0,0,0]
    for i in range(times):
        #palin
        lx,ly = x.tolist(), y.tolist()
        start = datetime.now()
        matrix_of_distance_plaint(lx,ly)
        res[0] += ((datetime.now() - start).total_seconds()/times)
        
        #numpy
        start = datetime.now()
        matrix_of_distance_numpy(x,y)
        res[1] += ((datetime.now() - start).total_seconds()/times)
        
        #scipy.spatial.distance.cdist
        start = datetime.now()
        scipy.spatial.distance.cdist(x,y)
        res[2] += ((datetime.now() - start).total_seconds()/times)
        
    measurements[1].append(res[0])
    measurements[2].append(res[1])
    measurements[3].append(res[2])
print(measurements)

fig = go.Figure()
fig.add_trace(go.Scatter(x=measurements[0], y=measurements[1],
                        mode='lines+markers',
                        name='palin python'))
fig.add_trace(go.Scatter(x=measurements[0], y=measurements[2],
                        mode='lines+markers',
                        name='numpy'))
fig.add_trace(go.Scatter(x=measurements[0], y=measurements[3],
                        mode='lines+markers', name='scipy'))

fig.show()


[[4, 5, 6, 7, 8, 9, 10], [3.215000000000003e-05, 3.067999999999999e-05, 5.2039999999999976e-05, 7.108000000000007e-05, 0.00010594999999999992, 9.097000000000005e-05, 0.00011320999999999997], [0.00021856000000000035, 0.0001242199999999999, 0.00023640000000000008, 0.00028779999999999957, 0.00046199000000000023, 0.00028562, 0.0003422300000000002], [2.1309999999999998e-05, 4.040000000000001e-06, 6.29e-06, 6.739999999999996e-06, 8.050000000000001e-06, 3.6800000000000033e-06, 3.910000000000003e-06]]


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [20]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [21]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [22]:
average_calories = calorie_stats.mean()
print(average_calories)

106.88311688311688


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [23]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [24]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 110 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [25]:

for i in range(100):
    v = np.percentile(calorie_stats,i)
    print(f'percentile({i}) = {v}')
    if v > 60:
        nth_percentile = i
        break


percentile(0) = 50.0
percentile(1) = 50.0
percentile(2) = 50.0
percentile(3) = 55.599999999999994
percentile(4) = 70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [26]:
more_calories = len(calorie_stats[calorie_stats > 60])/len(calorie_stats)
print(more_calories)

0.961038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [27]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

среднее у нас получилось ~ 106, а у нас колорий всего 60. Это больше чем два стандартных отклонения, такой важный факт можно использовать в маркетинге. Но важно помнить что калории это еще не все что нужно, нужно изучить какие калории у нас и у наших конкурентов, возможно у них там одны клетчатка, и усваиваемых каллорий у нас больше. 